python/dgl/core.py

```diff
@@ -255,6 +255,19 @@ def invoke_gspmm(graph, mfunc, rfunc, *, srcdata=None, dstdata=None, edata=None)
     else:
         x = alldata[mfunc.target][mfunc.in_field]
         op = getattr(ops, '{}_{}'.format(mfunc.name, rfunc.name))
+        if graph._graph.number_of_etypes() > 1:
+            # Convert to list as dict is unordered.
+            lhs_list = [None] * graph._graph.number_of_ntypes()
+            rhs_list = [None] * graph._graph.number_of_etypes()
+            if mfunc.name == "copy_u":
+                for srctype, _, _ in graph.canonical_etypes:
+                    src_id = graph.get_ntype_id(srctype)
+                    lhs_list[src_id] = x[srctype]
+            elif mfunc.name == "copy_e":
+                for rel in graph.canonical_etypes:
+                    etid = graph.get_etype_id(rel)
+                    rhs_list[etid] = x[rel]
+            x = tuple(lhs_list + rhs_list)
```

Comment:
Probably we want to add a utility function for this so later we could reuse it in many places. Sth like:
```python
def data_dict_to_tuple(data_dict, name, types):
    """Get node or edge feature data of the given name for all the types.

    Parameters
    -------------
    data_dict : dict[str, dict[str, Tensor]] or dict[str, dict[(str, str, str), Tensor]]
        Node or edge data stored in DGLGraph. The key of the outer dictionary is the feature name,
        while the key of the inner one is the node type name or edge type tuple.
    name : str
        Feature name.
    types : list[str] or list[(str, str, str)]
        Node type names or edge type tuples.

    Returns
    --------
    data_tuple : tuple(Tensor)
        Feature data stored in tuple of tensors. The i^th tensor stores the feature
        data of type ``types[i]``.
    """
    pass
```

For implementation, you could avoid calling `get_ntype_id` and `get_etype_id` because the type names you retrieved from DGLGraph are already in the type ID order. Simply use:
```python
for stid, srctype in enumerate(graph.srctypes):
    # stid is already source type ID
    ...

for etid, etype in enumerate(graph.canonical_etypes):
    # etid is already edge type ID
    ...
```

Reply:
I don't think the following two segments of code are same:
`for stid, srctype in enumerate(graph.srctypes):
    # stid is already source type ID
    ...`
 `for srctype, _, _ in graph.canonical_etypes:
       stid = graph.get_ntype_id(srctype)
       ...`
E.g., a graph has total #ntypes=3, #etypes=2 (ntype 0 -> ntype 1, ntype 2-> ntype 1). I want the srctypes to be 0 and 2. I think enumeration will always assign them as 0 and 1.

Reply:
I see. You are right.